src/message.rs

```diff
@@ -36,14 +36,160 @@ impl Deserialize for Version {
     }
 }
 
+/// An RPC method params
+///
+/// Not all json is a valid RPC params.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+#[serde(untagged)]
+pub enum Params {
+    Positional(Vec<Value>),
+    Named(serde_json::Map<String, Value>),
+}
+
+impl Params {
+    /// Wether or not there are any params
+    pub fn is_empty(&self) -> bool {
+        match *self {
+            Params::Positional(ref args) => args.is_empty(),
+            Params::Named(ref fields) => fields.is_empty(),
+        }
+    }
+
+    /// Returns a json value consuming the params in the process
+    pub fn into_value(self) -> Value {
+        match self {
+            Params::Positional(xs) => Value::Array(xs),
+            Params::Named(x) => Value::Object(x),
+        }
+    }
+
+    /// Force the params to be represented in positional form
+    pub fn into_positional(self) -> Self {
+        match self {
+            Params::Named(x) => {
+                let values = x.into_iter()
```

Comment:
There's a feature that forces the `Map` to preserve order (I don't remember the name), but enabling it would enable it for whatever other use the program might have for the `serde_json` crate. Otherwise, `Map` sorts them (that is why your tests with `{x: …, y: …}` didn't catch the problem, if you reverse them to `{y: …, x: …}`, then you might).

However, serde itself has two ways to decode structs, one is by a name and another is by the order, so it may be possible to place some kind of intermediate layer that converts it somehow. I don't know the details, only that some formats serde can decode/encode don't provide names, only the order (I think CBOR is one such format).